README.md
# Hyperscan

Hyperscan is a high-performance multiple regex matching library. It follows the
regular expression syntax of the commonly-used libpcre library, but is a
standalone library with its own C API.

Hyperscan uses hybrid automata techniques to allow simultaneous matching of
large numbers (up to tens of thousands) of regular expressions and for the
matching of regular expressions across streams of data.

Hyperscan is typically used in a DPI library stack.

# Documentation

Information on building the Hyperscan library and using its API is available in
the [Developer Reference Guide](http://intel.github.io/hyperscan/dev-reference/).

# License

Hyperscan is licensed under the BSD License. See the LICENSE file in the
project repository.

# Versioning

The `master` branch on Github/kunpengcompute will always contain the most recent 
release of Intel Hyperscan. 

The `aarch64` branch on Github/kunpengcompute will always contain the most recent 
release that supports the aarch64 architecture. The aarch64 branch was developed
based on Intel hyperscan 5.2.1. Each version released to `aarch64` branch goes through
QA and testing before it is released; if you're a user of aarch64, rather than a developer,
this is the version you should be using.

# Porting
Perform platform-specific different operations, including compilation options, 
detection specific header files, SIMD instruction judgment, and so on.

# Optimization
Improve the Kunpeng platform by using the NEON instructions, inline assembly, 
data alignment, instruction alignment, memory data prefetching, static branch 
prediction, code structure optimization, etc.

# Get Involved

The official homepage for Hyperscan is at [www.hyperscan.io](https://www.hyperscan.io).

`master` branch

If you have questions or comments, we encourage you to [join the mailing list]
(https://lists.01.org/mailman/listinfo/hyperscan). To file a bug, you can send an email 
to the list, or create an issue on Github.

If you wish to contact the Hyperscan team at Intel directly, without posting
publicly to the mailing list, send an email to
[hyperscan@intel.com](mailto:hyperscan@intel.com).

`aarch64` branch

If you have questions or comments, we encourage you to create an issue on Github.

If you wish to contact the Huawei team directly, you can send an email to 
kunpengcompute@huawei.com.


debian/changelog
hyperscan (5.3.0-1) unstable; urgency=medium

  * New upstream version 5.3.0
  * Move to debhelper-compat

 -- Robert Haist <rha@debian.org>  Wed, 10 Jun 2020 12:27:27 +0200

hyperscan (5.2.1-1) unstable; urgency=medium

  * New upstream version 5.2.1
  * Bump debian standards version to 4.4.1
  * Add Rules-Requires-Root

 -- Robert Haist <rhaist@mailbox.org>  Fri, 01 Nov 2019 13:11:11 +0100

hyperscan (5.2.0-2) unstable; urgency=medium

  * [a020a6d] Move to python3 (Closes: #936715)
  * [226de36] Bump debian standards version to 4.4.0
  * [35a20d1] Bump debhelper version to 12

 -- Robert Haist <rhaist@mailbox.org>  Fri, 06 Sep 2019 16:00:56 +0200

hyperscan (5.2.0-1) unstable; urgency=medium

  * [941c3fe] New upstream version 5.2.0

 -- Robert Haist <rhaist@mailbox.org>  Fri, 23 Aug 2019 15:15:24 +0200

hyperscan (5.1.1-1) unstable; urgency=medium

  * [7912722] New upstream version 5.1.1

 -- Robert Haist <rhaist@mailbox.org>  Mon, 22 Apr 2019 13:38:32 +0200

hyperscan (5.1.0-1) unstable; urgency=medium

  * New upstream version 5.1.0
  * Bump standards version
  * Remove duplicate example files
  * Add libpcap-dev build dependency for examples

 -- Robert Haist <rhaist@mailbox.org>  Thu, 31 Jan 2019 11:35:11 +0100

hyperscan (5.0.0-3) unstable; urgency=medium

  * Add Conflicts/Replaces so that the misnamed libhyperscan4 is removed
    on upgrades (Closes: #904571)

 -- Hilko Bengen <bengen@debian.org>  Thu, 26 Jul 2018 12:21:35 +0200

hyperscan (5.0.0-2) unstable; urgency=medium

  * Rename libhyperscan4 to libhyperscan5 (Closes: #903675)

 -- Robert Haist <rhaist@mailbox.org>  Sat, 21 Jul 2018 18:36:16 +0200

hyperscan (5.0.0-1) unstable; urgency=medium

  [ Sascha Steinbiss ]
  * add autopkgtest

  [ Robert Haist ]
  * Bump debhelper to compat level 11
  * Bump standards version
  * d/rules fix lintian debian-rules-uses-unnecessary-dh-argument
  * d/rules fix lintian debian-rules-is-dh_make-template
  * New upstream version 5.0.0

 -- Robert Haist <rhaist@mailbox.org>  Tue, 10 Jul 2018 10:54:02 +0200

hyperscan (4.7.0-1) unstable; urgency=medium

  * Add lintian-override for package-name-doesnt-match-sonames
  * Add Russian translation by Lev Lamberov (Closes: #883142)
  * d/compat bump to version 10
  * d/control change upstream github url
  * d/control increase debhelper dependency version
  * d/control change alioth to salsa repo
  * d/control bump standards version to 4.1.3
  * d/rules use dh_missing
  * New upstream version 4.7.0

 -- Robert Haist <rhaist@mailbox.org>  Wed, 24 Jan 2018 11:34:53 +0100

hyperscan (4.6.0-1) unstable; urgency=medium

  * New upstream version 4.6.0
  * Add AVX512 optimisations to fat runtime

 -- Robert Haist <rhaist@mailbox.org>  Sun, 24 Sep 2017 22:22:11 +0200

hyperscan (4.5.2-1) unstable; urgency=medium

  * New upstream version 4.5.2
  * Bump debian standards versionto 4.0.0

 -- Robert Haist <rhaist@mailbox.org>  Wed, 26 Jul 2017 20:55:04 +0200

hyperscan (4.5.1-1) unstable; urgency=medium

  [ Sascha Steinbiss ]
  * Non-maintainer upload.
  * Fix FTBFS on kFreeBSD.
  * Remove hurd-i386 from list of supported archs.

  [ Robert Haist ]
  * New upstream version 4.5.1
  * Remove patch because it went upstream

 -- Robert Haist <rhaist@mailbox.org>  Fri, 16 Jun 2017 21:32:48 +0200

hyperscan (4.5.0-1) unstable; urgency=medium

  * Add portuguese translations thanks to Rui Branco (Closes: #859334)
  * New upstream version 4.5.0
  * Fix multiarch directory changes and new example file location

 -- Robert Haist <rhaist@mailbox.org>  Mon, 12 Jun 2017 21:22:30 +0200

hyperscan (4.4.1-1) unstable; urgency=medium

  * Fix overlong changelog line to please lintian
  * Remove explicit FAT_RUNTIME cmake directive
  * New upstream version 4.4.1

 -- Robert Haist <rhaist@mailbox.org>  Sat, 04 Mar 2017 17:06:09 +0100

hyperscan (4.4.0-1) unstable; urgency=medium

  [ Robert Haist ]
  * Add brazilian / portuguese translations thanks to Adriano Rafael Gomes
    (Closes: #846528)
  * New upstream version 4.4.0

  [ Sascha Steinbiss ]
  * use secure Vcs-Git
  * add missing build-deps

 -- Robert Haist <rhaist@mailbox.org>  Fri, 20 Jan 2017 10:08:50 +0100

hyperscan (4.3.1-1) unstable; urgency=medium

  [ Hilko Bengen ]
  * Add gbp.conf

  [ Robert Haist ]
  * New upstream version 4.3.1

 -- Robert Haist <rhaist@mailbox.org>  Mon, 12 Sep 2016 22:28:09 +0200

hyperscan (4.3.0-1) unstable; urgency=medium

  [ Robert Haist ]
  * Set correct Vcs-Browser link
  * Added Dutch translations for debconf provided by Frans Spiesschaert
    (Closes: #834623)

  [ Hilko Bengen ]
  * New upstream version
  * Drop reproducible timestamp patch that was integrated upstream
  * Rename library package to libhyperscan4

 -- Hilko Bengen <bengen@debian.org>  Wed, 24 Aug 2016 23:09:17 +0200

hyperscan (4.2.0-3) unstable; urgency=medium

  [ Robert Haist ]
  * Fix build errors for gcc 6.1.1 by explicitly setting CMAKE_BUILD_TYPE
  * Point Vcs-Git & Vcs-Browser to the src pkg git and not to upstream

  [ SZ Lin ]
  * Added zh_TW translations for debconf

  [ Hilko Bengen ]
  * Update VCS links to anonscm.debian.org
  * Add French debconf translation provided by Julien Patriarca (Closes:
    #833855)
  * Add German debconf translation
  * Add myself to Uploaders

 -- Hilko Bengen <bengen@debian.org>  Tue, 09 Aug 2016 22:10:49 +0200

hyperscan (4.2.0-2) unstable; urgency=medium

  [ Eduard Sanou ]
  * Use SOURCE_DATE_EPOCH as embedded timestamp to make the build
    reproducible (Closes: #832917).

  [ Robert Haist ]
  * Fix builds (E-march-native) for i386, kfreebsd-amd64, kfreebsd-i386, x32.

 -- Robert Haist <rhaist@mailbox.org>  Fri, 29 Jul 2016 20:49:09 +0200

hyperscan (4.2.0-1) unstable; urgency=medium

  * Initial release (Closes: #824838)

 -- Robert Haist <rhaist@mailbox.org>  Mon, 25 Jul 2016 16:44:23 +0200


src/util/cpuid_inline.h
/*
 * Copyright (c) 2017, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Intel Corporation nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef CPUID_INLINE_H_
#define CPUID_INLINE_H_

#include "ue2common.h"
#include "cpuid_flags.h"

#if !defined(_WIN32) && !defined(CPUID_H_)
#include <cpuid.h>
/* system header doesn't have a header guard */
#define CPUID_H_
#endif

#ifdef __cplusplus
extern "C"
{
#endif

static inline
void cpuid(unsigned int op, unsigned int leaf, unsigned int *eax,
           unsigned int *ebx, unsigned int *ecx, unsigned int *edx) {
#ifndef _WIN32
    //__cpuid_count(op, leaf, *eax, *ebx, *ecx, *edx);
#else
    int a[4];
    __cpuidex(a, op, leaf);
    *eax = a[0];
    *ebx = a[1];
    *ecx = a[2];
    *edx = a[3];
#endif
}

// ECX
#define CPUID_SSE3 (1 << 0)
#define CPUID_SSSE3 (1 << 9)
#define CPUID_SSE4_1 (1 << 19)
#define CPUID_SSE4_2 (1 << 20)
#define CPUID_POPCNT (1 << 23)
#define CPUID_XSAVE (1 << 27)
#define CPUID_AVX (1 << 28)

// EDX
#define CPUID_FXSAVE (1 << 24)
#define CPUID_SSE (1 << 25)
#define CPUID_SSE2 (1 << 26)
#define CPUID_HTT (1 << 28)

// Structured Extended Feature Flags Enumeration Leaf ECX values
#define CPUID_BMI (1 << 3)
#define CPUID_AVX2 (1 << 5)
#define CPUID_BMI2 (1 << 8)

// Structured Extended Feature Flags Enumeration Leaf EBX values
#define CPUID_AVX512F (1 << 16)
#define CPUID_AVX512BW (1 << 30)

// Extended Control Register 0 (XCR0) values
#define CPUID_XCR0_SSE (1 << 1)
#define CPUID_XCR0_AVX (1 << 2)
#define CPUID_XCR0_OPMASK (1 << 5) // k-regs
#define CPUID_XCR0_ZMM_Hi256 (1 << 6) // upper 256 bits of ZMM0-ZMM15
#define CPUID_XCR0_Hi16_ZMM (1 << 7) // ZMM16-ZMM31

#define CPUID_XCR0_AVX512                                                      \
    (CPUID_XCR0_OPMASK | CPUID_XCR0_ZMM_Hi256 | CPUID_XCR0_Hi16_ZMM)

static inline
u64a xgetbv(u32 op) {
#if defined(_WIN32) || defined(__INTEL_COMPILER)
    return _xgetbv(op);
#else
    u32 a, d;
    //__asm__ volatile (
    //        "xgetbv\n"
    //        : "=a"(a),
    //          "=d"(d)
    //        : "c"(op));
    return ((u64a)d << 32) + a;
#endif
}

static inline
int check_avx2(void) {
#if defined(__INTEL_COMPILER)
    return _may_i_use_cpu_feature(_FEATURE_AVX2);
#else
    unsigned int eax, ebx, ecx, edx;

    cpuid(1, 0, &eax, &ebx, &ecx, &edx);

    /* check AVX is supported and XGETBV is enabled by OS */
    if ((ecx & (CPUID_AVX | CPUID_XSAVE)) != (CPUID_AVX | CPUID_XSAVE)) {
        DEBUG_PRINTF("AVX and XSAVE not supported\n");
        return 0;
    }

    /* check that SSE and AVX registers are enabled by OS */
    u64a xcr0 = xgetbv(0);
    if ((xcr0 & (CPUID_XCR0_SSE | CPUID_XCR0_AVX)) !=
        (CPUID_XCR0_SSE | CPUID_XCR0_AVX)) {
        DEBUG_PRINTF("SSE and AVX registers not enabled\n");
        return 0;
    }

    /* ECX and EDX contain capability flags */
    ecx = 0;
    cpuid(7, 0, &eax, &ebx, &ecx, &edx);

    if (ebx & CPUID_AVX2) {
        DEBUG_PRINTF("AVX2 enabled\n");
        return 1;
    }

    return 0;
#endif
}

static inline
int check_avx512(void) {
    /*
     * For our purposes, having avx512 really means "can we use AVX512BW?"
     */
#if defined(__INTEL_COMPILER)
    return _may_i_use_cpu_feature(_FEATURE_AVX512BW | _FEATURE_AVX512VL);
#else
    unsigned int eax, ebx, ecx, edx;

    cpuid(1, 0, &eax, &ebx, &ecx, &edx);

    /* check XSAVE is enabled by OS */
    if (!(ecx & CPUID_XSAVE)) {
        DEBUG_PRINTF("AVX and XSAVE not supported\n");
        return 0;
    }

    /* check that AVX 512 registers are enabled by OS */
    u64a xcr0 = xgetbv(0);
    if ((xcr0 & CPUID_XCR0_AVX512) != CPUID_XCR0_AVX512) {
        DEBUG_PRINTF("AVX512 registers not enabled\n");
        return 0;
    }

    /* ECX and EDX contain capability flags */
    ecx = 0;
    cpuid(7, 0, &eax, &ebx, &ecx, &edx);

    if (!(ebx & CPUID_AVX512F)) {
        DEBUG_PRINTF("AVX512F (AVX512 Foundation) instructions not enabled\n");
        return 0;
    }

    if (ebx & CPUID_AVX512BW) {
        DEBUG_PRINTF("AVX512BW instructions enabled\n");
        return 1;
    }

    return 0;
#endif
}

static inline
int check_ssse3(void) {
//    unsigned int eax, ebx, ecx, edx;
//    cpuid(1, 0, &eax, &ebx, &ecx, &edx);
//    return !!(ecx & CPUID_SSSE3);
      return 1;
}

static inline
int check_sse42(void) {
    unsigned int eax, ebx, ecx, edx;
    cpuid(1, 0, &eax, &ebx, &ecx, &edx);
    return !!(ecx & CPUID_SSE4_2);
}

static inline
int check_popcnt(void) {
    unsigned int eax, ebx, ecx, edx;
    cpuid(1, 0, &eax, &ebx, &ecx, &edx);
    return !!(ecx & CPUID_POPCNT);
}

#ifdef __cplusplus
} /* extern "C" */
#endif

#endif /* CPUID_INLINE_H_ */
